### PR TITLE
Shutdown TimeoutManager during node Shutdown

### PR DESCRIPTION
**Pull request by abi87**
## Why this should be merged
@darioush introduced goleak to [fix some goroutines leak in coreth UTs](https://github.com/ava-labs/coreth/pull/273). I applied it to some avalanchego packages and came up with this cleanup. I believe we are leaking some goroutines in production code too.

## How this works
Just add a main_test.go in a few package and run UTs as usual. An error log comes up if goroutines leak. I fix them, mostly just duly called a shutdown method wherever needed.

## How this was tested
Fixes mainly UTs + timeoutManager in prod code

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -137,6 +137,7 @@ require (
 	go.opentelemetry.io/otel/exporters/otlp/internal/retry v1.11.0 // indirect
 	go.opentelemetry.io/proto/otlp v0.19.0 // indirect
 	go.uber.org/atomic v1.10.0 // indirect
+	go.uber.org/goleak v1.2.1 // indirect
```
Comment (darioush) [inline go.mod:140]: should be direct, could we run go mod tidy

Comment (abi87) [reply go.mod]: Done

File: vms/platformvm/service_test.go [hunk 2]
```diff
@@ -724,7 +724,10 @@ func TestGetBlock(t *testing.T) {
 			require := require.New(t)
 			service, _ := defaultService(t)
 			service.vm.ctx.Lock.Lock()
-			defer service.vm.ctx.Lock.Unlock()
+			defer func() {
+				require.NoError(service.vm.Shutdown(context.Background()))
+				service.vm.ctx.Lock.Unlock()
+			}()
```
Comment (abi87) [inline vms/platformvm/service_test.go:730]: avoids leaking timer goroutine, similarly to other tests

File: vms/platformvm/vm_test.go [hunk 3]
```diff
@@ -1561,6 +1561,7 @@ func TestBootstrapPartiallyAccepted(t *testing.T) {
 	require.NoError(err)
 
 	go timeoutManager.Dispatch()
+	defer timeoutManager.Stop()
```
Comment (abi87) [inline vms/platformvm/vm_test.go:1564]: avoids leaking timer goroutine

File: node/node.go [hunk 4]
```diff
@@ -125,6 +125,9 @@ type Node struct {
 	// Build and parse messages, for both network layer and chain manager
 	msgCreator message.Creator
 
+	// Manages network timeouts
+	timeoutManager timeout.Manager
```
Comment (abi87) [inline node/node.go:129]: making this a node attribute to duly stop it upon shutdown.

File: snow/networking/timeout/manager_test.go [hunk 5]
```diff
@@ -33,17 +32,13 @@ func TestManagerFire(t *testing.T) {
 	)
 	require.NoError(t, err)
 	go manager.Dispatch()
-
-	wg := sync.WaitGroup{}
-	wg.Add(1)
```
Comment (abi87) [inline snow/networking/timeout/manager_test.go:38]: this still leaks the goroutine in manager.Dispatch

**Review by darioush — APPROVED** [review]

File: snow/networking/timeout/manager_test.go [hunk 6]
```diff
@@ -33,17 +32,13 @@ func TestManagerFire(t *testing.T) {
...
 	manager.RegisterRequest(
 		ids.NodeID{},
 		ids.ID{},
 		true,
 		ids.RequestID{},
-		wg.Done,
+		nil,
 	)
-
-	wg.Wait()
```
Comment (StephenButtolph) [inline snow/networking/timeout/manager_test.go:48]: This seems to be deleting the actual test

Comment (abi87) [reply snow/networking/timeout/manager_test.go]: Indeed I misunderstood the point of the test. Fixed.

File: snow/networking/timeout/manager.go [hunk 7]
```diff
@@ -153,3 +158,9 @@ func (m *manager) RemoveRequest(requestID ids.RequestID) {
 func (m *manager) RegisterRequestToUnreachableValidator() {
 	m.tm.ObserveLatency(m.TimeoutDuration())
 }
+
+func (m *manager) Stop() {
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply snow/networking/timeout/manager.go:165]: I think doing this change makes the PR bigger so going to approve and punt this off for later

File: snow/networking/router/main_test.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+package router
+
+import (
+	"testing"
+
+	"go.uber.org/goleak"
+)
+
+// TestMain uses goleak to verify tests in this package do not leak unexpected
+// goroutines.
```
Comment (joshua-kim) [inline snow/networking/router/main_test.go:13]: nit: if the code is self-documenting I usually don't put a comment on it, i would prefer to just delete this comment

File: snow/networking/timeout/main_test.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+package timeout
+
+import (
+	"testing"
+
+	"go.uber.org/goleak"
+)
+
+// TestMain uses goleak to verify tests in this package do not leak unexpected
+// goroutines.
```
Comment (joshua-kim) [inline snow/networking/timeout/main_test.go:13]: Same comment here

File: vms/platformvm/block/builder/helpers_test.go [hunk 10]
```diff
@@ -409,6 +409,8 @@ func buildGenesisTest(t *testing.T, ctx *snow.Context) []byte {
 }
 
 func shutdownEnvironment(env *environment) error {
+	env.Builder.Shutdown() // avoids leaking timer goroutine
```
Comment (joshua-kim) [inline vms/platformvm/block/builder/helpers_test.go:412]: nit: remove comment

File: vms/platformvm/block/builder/main_test.go [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+package builder
+
+import (
+	"testing"
+
+	"go.uber.org/goleak"
+)
+
+// TestMain uses goleak to verify tests in this package do not leak unexpected
+// goroutines.
```
Comment (joshua-kim) [inline vms/platformvm/block/builder/main_test.go:13]: Same comment

File: vms/platformvm/main_test.go [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+package platformvm
+
+import (
+	"testing"
+
+	"go.uber.org/goleak"
+)
+
+// TestMain uses goleak to verify tests in this package do not leak unexpected
+// goroutines.
```
Comment (joshua-kim) [inline vms/platformvm/main_test.go:13]: See earlier comment

File: snow/networking/router/chain_router_test.go [hunk 13]
```diff
@@ -1526,7 +1578,9 @@ func TestValidatorOnlyAllowedNodeMessageDrops(t *testing.T) {
 		prometheus.NewRegistry(),
 	)
 	require.NoError(err)
+
```
Comment (joshua-kim) [inline snow/networking/router/chain_router_test.go:1581]: nit: clear out these new line diffs

File: snow/networking/router/chain_router_test.go [hunk 14]
```diff
@@ -711,12 +725,15 @@ func TestRouterHonorsRequestedEngine(t *testing.T) {
 		"",
 		prometheus.NewRegistry(),
 	))
+	defer chainRouter.Shutdown(context.TODO())
```
Comment (joshua-kim) [inline snow/networking/router/chain_router_test.go:728]: We should use `context.Background` in tests instead of `TODO`

**Review by joshua-kim — COMMENTED** [review]
Last few nits to remove the new line diffs + remove some of the unnecessary comments

File: snow/networking/router/chain_router_test.go (hunk 14 above)
Comment (abi87) [reply snow/networking/router/chain_router_test.go:728]: Done

File: snow/networking/router/chain_router_test.go (hunk 13 above)
Comment (abi87) [reply snow/networking/router/chain_router_test.go:1581]: @joshua-kim , If you see above I grouped
```
go tm.Dispatch()
defer tm.Stop()
```
in all tests. Should I revert the lines in all tests?

File: snow/networking/router/chain_router_test.go (hunk 14 above)
Comment (abi87) [reply snow/networking/router/chain_router_test.go:728]: Done

File: snow/networking/router/main_test.go (hunk 8 above)
Comment (abi87) [reply snow/networking/router/main_test.go:13]: Dropped all of them

File: vms/platformvm/block/builder/helpers_test.go (hunk 10 above)
Comment (abi87) [reply vms/platformvm/block/builder/helpers_test.go:412]: Done

File: snow/networking/router/chain_router_test.go [hunk 15]
```diff
@@ -325,6 +329,7 @@ func TestShutdownTimesOut(t *testing.T) {
 // Ensure that a timeout fires if we don't get a response to a request
 func TestRouterTimeout(t *testing.T) {
 	require := require.New(t)
+
```
Comment (abi87) [inline snow/networking/router/chain_router_test.go:332]: all the test have require in its own "block" separated by a line from the rest of the test

File: snow/networking/router/chain_router_test.go (hunk 13 above)
Comment (joshua-kim) [reply snow/networking/router/chain_router_test.go:1581]: Ah I see

**Review by joshua-kim — APPROVED** [review]

File: snow/networking/router/chain_router_test.go [hunk 16]
```diff
@@ -1543,6 +1597,7 @@ func TestValidatorOnlyAllowedNodeMessageDrops(t *testing.T) {
 		"",
 		prometheus.NewRegistry(),
 	))
+	defer chainRouter.Shutdown(context.TODO())
```
Comment (StephenButtolph) [inline snow/networking/router/chain_router_test.go:1600]: I don't think we're ever planning on having a more canonical context in the tests - so normally we just use `Background`.
```suggestion
	defer chainRouter.Shutdown(context.Background())
```
(I think this applies a few times to this file)

**Review by StephenButtolph — APPROVED** [review]

**Pull request merged**